core/Network/TLS/Handshake/Server.hs

```diff
@@ -315,9 +315,24 @@ doHandshake sparams mcred ctx chosenVersion usedCipher usedCompression clientSes
             usingHState ctx $ setDHPrivate priv
             return serverParams
 
+        decideHash sigAlg = do
+            usedVersion <- usingState_ ctx getVersion
+            case usedVersion of
+              TLS12 -> do
+                  let sHashSigs = supportedHashSignatures $ ctxSupported ctx
+                      cHashSigs = case extensionLookup extensionID_SignatureAlgorithms exts >>= extensionDecode False of
+                          Just (SignatureAlgorithms hss) -> hss
+                          Nothing                        -> []
+                      hashSigs = cHashSigs `intersect` sHashSigs
+                  case filter ((==) sigAlg . snd) hashSigs of
+                      []  -> error ("no hash signature for " ++ show sigAlg)
+                      x:_ -> return $ Just (fst x)
```

Comment:
A server already chooses one cipher suite from a list of cipher suites.
The signature algorithm is decided from the chosen cipher.
A missing thing here is a hash algorithm which is not indicated by the cipher suite.
In TLS 1.2, the hash algorithm should be chosen from the signature algorithm extensions, if provided.
This hash is used only to sign (EC)DHE parameters, not RSA stuff.
So, I believe that no inconsistency for hash happens.